src/test/kotlin/org/elm/lang/core/completion/ElmRecordExprCompletionTest.kt

```diff
@@ -0,0 +1,193 @@
+package org.elm.lang.core.completion
+
+class ElmRecordExprCompletionTest : ElmCompletionTestBase() {
+
+    fun `test blank`() = doSingleCompletion(
+            """
+f : { foo : () }
+f =
+    { {-caret-} }
+""", """
+f : { foo : () }
+f =
+    { foo = {-caret-} }
+""")
+
+    fun `test one letter`() = doSingleCompletion(
+            """
+f : { foo : (), other: () }
+f =
+    { f{-caret-} }
+""", """
+f : { foo : (), other: () }
+f =
+    { foo = {-caret-} }
+""")
+
+    fun `test no completions when all fields already defined`() = checkNoCompletion(
+            """
+f : { foo : () }
+f =
+    { foo = "", f{-caret-} }
+""")
+
+    fun `test no completions when type cannot be inferred`() = checkNoCompletion(
+            """
+f =
+    { foo = "", f{-caret-} }
+""")
+
+    fun `test middle of record missing trailing comma`() = doSingleCompletion(
+            """
+f : { foo : (), bar : (), baz : () }
+f =
+    { foo = ()
+    , b{-caret-}
+      baz = ()
+    }
+""", """
+f : { foo : (), bar : (), baz : () }
+f =
+    { foo = ()
+    , bar = {-caret-}
+      baz = ()
+    }
+""")
+
+    fun `test record update`() = doSingleCompletion(
+            """
+type alias Foo = { foo : () }
+
+f : Foo -> Foo
+f foo =
+    { foo | f{-caret-} }
+""", """
+type alias Foo = { foo : () }
+
+f : Foo -> Foo
+f foo =
+    { foo | foo = {-caret-} }
+""")
+
+    fun `test root of nested record`() = doSingleCompletion(
+            """
+f : { foo : { bar: () } }
+f =
+    { f{-caret-} }
+""", """
+f : { foo : { bar: () } }
+f =
+    { foo = {-caret-} }
+""")
+
+    fun `test leaf of nested record`() = checkNoCompletion(
```

Comment:
Is there a typo here? Should the s{-caret-} be changed to b{-caret-} to match the first letter of bar?

If this is a typo how come completions do not work inside nested records?

If this is not a typo then consider updating the name of the test.

Reply:
I don't even know why that test was there. 

The reason nested records don't work is the same reason that tuples didn't: we report a single error on the entire outer record rather than more granular errors. 

The fix would be the same as the fix for tuples: report errors on the individual fields. That would work for extra or mismatched fields, but not for missing. I don't know of a good way to report missing fields other than what we do now.

Reply:
I ran into the same issue in my attempt. I started playing around with showing the more granular errors when all the expected keys in the outer record were present otherwise showing the error for the entire outer record. Do you think something like that could make sense?

Reply:
That's basically what I'm doing with tuples. I think it might feel inconsistent to do the same thing with records, but maybe it would be ok.